stellargraph/mapper/graphwave_generator.py

```diff
@@ -0,0 +1,176 @@
+        self.node_list = G.nodes_of_type(node_types[0])
+        self.Aadj = G.to_adjacency_matrix(self.node_list)
+
+        # Function to map node IDs to indices for quicker node index lookups
+        # TODO: Move this to the graph class
+        node_index_dict = dict(zip(self.node_list, range(len(self.node_list))))
+        self._node_lookup = np.vectorize(node_index_dict.get, otypes=[np.int64])
+
+        adj = G.to_adjacency_matrix().tocoo()
+        degree_mat = diags(np.array(adj.sum(1)).flatten())
+        laplacian = degree_mat - adj
+
+        if num_eigenvecs == -1:
+            num_eigenvecs = laplacian.shape[0] - 2
+
+        self.eigen_vals, self.eigen_vecs = eigs(laplacian, k=num_eigenvecs)
+        self.eigen_vals = np.real(self.eigen_vals).astype(np.float32)
+        self.eigen_vecs = np.real(self.eigen_vecs).astype(np.float32)
+
+        if scales == "auto":
+
+            e2 = self.eigen_vals[self.eigen_vals > 0].min()
+            eN = self.eigen_vals.max()
+
+            min_scale = -np.log(0.95) / np.sqrt(eN * e2)
+            max_scale = -np.log(0.85) / np.sqrt(eN * e2)
+
+            scales = np.linspace(min_scale, max_scale, num_scales)
+
+        self.scales = scales
+
+        # the columns of U exp(-scale * eigenvalues) U^T (U = eigenvectors) are used to calculate the node embeddings
+        # (each column corresponds to a node)
+        # to avoid computing a dense NxN matrix when only several eigenvalues are specified
+        # U exp(-scale * eigenvalues) is computed and stored - which is an N x num_eigenvectors matrix
+        # the columns of exp(-scale * eigenvalues) U^T are then computed on the fly in generator.flow()
+        self.Ues = [
+            self.eigen_vecs.dot(np.diag(np.exp(-s * self.eigen_vals))) for s in scales
+        ]  # a list of [U exp(-scale * eigenvalues) for scale in scales]
+        self.Ues = tf.convert_to_tensor(np.dstack(self.Ues))
+
+    def flow(
+        self, node_ids, sample_points, batch_size, targets=None, repeat=True, threads=1
+    ):
+        """
+        Creates a tensorflow DataSet object of GraphWave embeddings.
+
+        Args:
+            node_ids: an iterable of node ids for the nodes of interest
+                (e.g., training, validation, or test set nodes)
+            sample_points: a 1D array of points at which to sample the characteristic function.
+            batch_size: the number of node embeddings to include in a batch.
+            targets: a 1D or 2D array of numeric node targets with shape `(len(node_ids)`
+                or (len(node_ids), target_size)`
+            repeat (bool): indicates whether iterating through the DataSet will continue infinitely or stop after one
+                full pass.
+            threads (int): number of threads to use.
+        """
+        ts = tf.convert_to_tensor(sample_points.astype(np.float32))
+
+        dataset = (
+            tf.data.Dataset.from_tensor_slices(
```

Comment:
I was avoiding calculating `tf.einsum("ijk,jl->ilk", self.Ues, rows)` explicitly because its an `N x N x scales` size matrix. By breaking up the calculation by rows we should be able to handle much larger graphs.